sycl/source/detail/device_kernel_info.cpp

```diff
@@ -0,0 +1,89 @@
+//==---------------------- device_kernel_info.cpp ----------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#include <detail/device_kernel_info.hpp>
+#include <detail/program_manager/program_manager.hpp>
+
+namespace sycl {
+inline namespace _V1 {
+namespace detail {
+
+DeviceKernelInfo::DeviceKernelInfo(const CompileTimeKernelInfoTy &Info)
+    : CompileTimeKernelInfoTy(Info)
+#ifndef __INTEL_PREVIEW_BREAKING_CHANGES
+      ,
+      Name(Info.Name.data())
+#endif
+{
+  init(Name.data());
+}
+
+void DeviceKernelInfo::init(KernelNameStrRefT KernelName) {
+  auto &PM = detail::ProgramManager::getInstance();
+  MUsesAssert = PM.kernelUsesAssert(KernelName);
+  MImplicitLocalArgPos = PM.kernelImplicitLocalArgPos(KernelName);
+#ifndef __INTEL_PREVIEW_BREAKING_CHANGES
+  MInitialized.store(true);
+#endif
+}
+
+#ifndef __INTEL_PREVIEW_BREAKING_CHANGES
+void DeviceKernelInfo::initIfNeeded(KernelNameStrRefT KernelName) {
+  if (!MInitialized.load())
+    init(KernelName);
+}
+#endif
+
+template <typename OtherTy>
+inline constexpr bool operator==(const CompileTimeKernelInfoTy &LHS,
+                                 const OtherTy &RHS) {
+  // TODO replace with std::tie(...) == std::tie(...) once there is
+  // implicit conversion from detail to std string_view.
+  return std::string_view{LHS.Name} == std::string_view{RHS.Name} &&
+         LHS.NumParams == RHS.NumParams && LHS.IsESIMD == RHS.IsESIMD &&
+         std::string_view{LHS.FileName} == std::string_view{RHS.FileName} &&
+         std::string_view{LHS.FunctionName} ==
+             std::string_view{RHS.FunctionName} &&
+         LHS.LineNumber == RHS.LineNumber &&
+         LHS.ColumnNumber == RHS.ColumnNumber &&
+         LHS.KernelSize == RHS.KernelSize &&
+         LHS.ParamDescGetter == RHS.ParamDescGetter &&
+         LHS.HasSpecialCaptures == RHS.HasSpecialCaptures;
+}
+
+void DeviceKernelInfo::setCompileTimeInfoIfNeeded(
+    const CompileTimeKernelInfoTy &Info) {
+  if (isCompileTimeInfoSet())
+    CompileTimeKernelInfoTy::operator=(Info);
+  assert(isCompileTimeInfoSet());
```

Comment:
@sergey-semenov , what did you mean by those lines?

Reply:
https://github.com/intel/llvm/pull/20003

Reply:
> Why isn't it failing anywhere?

That probably indicates lack of test coverage for this scenario. If that's true, I don't think it's worth adding since my next PR will essentially make all kernel info behave like this (i.e. runtime information will be filled out during image registration, compile time information will be filled out during the first access to device kernel info with compile time information available).

>When I tried the change (although on top of my other changes), things broke for me. I'm not sure if it'd be clean on trunk either. Will you follow up please?

I'll look into it, let's see if https://github.com/intel/llvm/pull/20003 hits any failures in precommit.